### PR TITLE
Testcase and helpers.py enhancement

### PR DESCRIPTION
**Pull request by shreekarSS**
tests/helpers.py
  - Changed create_pvc() to instanciate PVC
  - Modifications in create_pod()
  - Added 'interface' to resource name while creating resource
  - Added new functions
     -- get_all_pvs() --> Returns all pvc in given namespace
     -- validate_pv_delete() --> Validates if pv is deleted after pvc deletion with retry()

tests/manage/test_create_storage_class_pvc.py
 - Updated fixtures for cephfs and rbd
 - Added validate_pv_delete() for validating pv after pvc delete
 - Added pod creation for cephfs and rbd with default values

ocs_ci/ocs/constants.py
 - Added rbd and cephfs interfaces

Fixes: https://github.com/red-hat-storage/ocs-ci/issues/343

### CODE REVIEW

**Comment by shreekarSS** [comment]
Same PR as https://github.com/red-hat-storage/ocs-ci/pull/332, was closed by mistake

File: tests/helpers.py [hunk 1]
```diff
@@ -56,41 +54,66 @@ def create_resource(
...
+            condition=state, resource_name=resource.name
+        )
+    except TimeoutExpiredError:
+        logging.warning(f'Failed to reach {resource.kind} {resource.name} to desired {state}')
+        return False
+    logger.info(f"{resource.kind} {resource.name} reached state {state}")
+    return True
+
+
+def create_pod(interface_type=None, pvc=None):
```
Comment (ebenahar) [inline tests/helpers.py]: The change in this function will affect its calls. For example, in tests/fixtures/create_pod()

**Comment by shreekarSS** [comment]
result link http://pastebin.test.redhat.com/777626

File: tests/helpers.py [hunk 2]
```diff
@@ -401,12 +425,9 @@ def delete_cephblockpool(pool_name=None):
         namespace=defaults.ROOK_CLUSTER_NAMESPACE
     )
     pool_list = get_cephblockpool_name()
-    if pool_name:
-        assert POOL.delete(resource_name=pool_name)
-    else:
-        for item in pool_list:
-            logger.info(f"Deleting CephBlockPool with name {item}")
-            assert POOL.delete(resource_name=item)
+    for item in pool_list:
```
Comment (RazTamir) [inline tests/helpers.py]: are you deleting all cephblockpools?

File: tests/helpers.py [hunk 3]
```diff
@@ -33,16 +36,11 @@ def create_unique_resource_name(resource_description, resource_type):
     return f"{resource_type}-{resource_description[:23]}-{current_date_time[:10]}"
 
 
-def create_resource(
-    desired_status=constants.STATUS_AVAILABLE, wait=True, **kwargs
-):
+def create_resource(**kwargs):
```
Comment (RazTamir) [inline tests/helpers.py]: please ensure all calls to his func has changed as you removed the parameters it gets

File: tests/helpers.py [hunk 4]
```diff
@@ -527,3 +543,37 @@ def run_io_with_rados_bench(**kw):
...
+        pv (str): pv from pvc to validates
+    Returns:
+        bool: True if deletion is successful
+    """
+    ocp_pv_list = get_all_pvs()
+    pv_all_list = ocp_pv_list['items']
+    if pv_all_list:
+        for item in pv_all_list:
+            if item['metadata']['name'] == pv:
+                raise AssertionError
```
Comment (RazTamir) [inline tests/helpers.py]: the docstring says this function return bool

File: tests/helpers.py [hunk 5]
```diff
@@ -527,3 +543,37 @@ def run_io_with_rados_bench(**kw):
...
+def validate_pv_delete(pv):
+    """
+    validates if pv is deleted after pvc deletion
+
+    Args:
+        pv (str): pv from pvc to validates
+    Returns:
+        bool: True if deletion is successful
+    """
+    ocp_pv_list = get_all_pvs()
```
Comment (RazTamir) [inline tests/helpers.py]: why you are getting all pvs and check if one is equal to 'pv'?
just try to .get(resource_name=pv)

File: tests/manage/test_create_storage_class_pvc.py [hunk 6]
```diff
@@ -1,135 +1,139 @@
...
-        self.cephfs_secret['data']['adminKey'] = (
-            get_admin_key_from_ceph_tools()
-        )
-        self.cephfs_secret['data']['adminID'] = constants.ADMIN_BASE64
-        logging.info(self.cephfs_secret)
-        secret = OCS(**self.cephfs_secret)
-        secret.create()
-        self.cephfs_sc = templating.load_yaml_to_dict(
-            constants.CSI_CEPHFS_STORAGECLASS_YAML
+        global RBD_PVC_OBJ, RBD_POD_OBJ, rbd_pv
```
Comment (RazTamir) [inline tests/manage/test_create_storage_class_pvc.py]: rbd_pv is global?
why not in a capital letters

File: tests/manage/test_create_storage_class_pvc.py [hunk 7]
```diff
@@ -1,135 +1,139 @@
...
         )
-        self.cephfs_sc['parameters']['monitors'] = self.mons
-        self.cephfs_sc['parameters']['pool'] = (
-            f"{self.fs_data['metadata']['name']}-data0"
+        RBD_PVC_OBJ = helpers.create_pvc(RBD_SC_OBJ.name, pvc_name)
+        ocp_pv_list = helpers.get_all_pvs()
+        pv_list = ocp_pv_list['items']
+        for pv in pv_list:
+            if pv['spec']['claimRef']['name'] == RBD_PVC_OBJ.name:
+                rbd_pv = pv['metadata']['name']
```
Comment (RazTamir) [inline tests/manage/test_create_storage_class_pvc.py]: what are you trying to do here?

File: tests/manage/test_create_storage_class_pvc.py [hunk 8]
```diff
@@ -1,135 +1,139 @@
...
-        del self.rbd_secret['data']['kubernetes']
-        self.rbd_secret['data']['admin'] = get_admin_key_from_ceph_tools()
-        logging.info(self.rbd_secret)
-        secret = OCS(**self.rbd_secret)
-        secret.create()
-        self.rbd_sc = templating.load_yaml_to_dict(
-            constants.CSI_RBD_STORAGECLASS_YAML
+        ocp_pv_list = helpers.get_all_pvs()
+        pv_list = ocp_pv_list['items']
+        for pv in pv_list:
```
Comment (RazTamir) [inline tests/manage/test_create_storage_class_pvc.py]: same question - what are you trying to do here?

File: tests/helpers.py [hunk 9]
```diff
@@ -155,8 +180,7 @@ def create_ceph_block_pool(pool_name=None):
 
 
 def create_storage_class(
```
Comment (prsurve) [inline tests/helpers.py]: missing some parameters from #337 pr plz fix this might fail tests/manage/test_verify_rbd_based_rwo_pvc.py script

File: tests/helpers.py [hunk 10]
```diff
@@ -240,9 +264,10 @@ def create_pvc(sc_name, pvc_name=None):
     )
     pvc_data['metadata']['namespace'] = defaults.ROOK_CLUSTER_NAMESPACE
     pvc_data['spec']['storageClassName'] = sc_name
-    return create_resource(
-        desired_status=constants.STATUS_BOUND, **pvc_data
-    )
+    ocs_obj = pvc.PVC(**pvc_data)
+    created_pvc = ocs_obj.create()
+    assert created_pvc, f"Failed to create resource {pvc_name}"
```
Comment (prsurve) [inline tests/helpers.py]: This will not fail if pvc is in Pending don't we need to check status of pvc if it's bound or pending

**Comment by vasukulkarni** [comment]
This is a same PR as #332 and I hope it hasn't changed from what was in the original PR, @shreekarSS can you address those comments fast, I would like to see it get merged fast so that we can see >90% PASS RATE for tier1.

File: tests/helpers.py (hunk 2 above)
Comment (shreekarSS) [reply tests/helpers.py]: No, It needs rebase with latest master

File: tests/helpers.py (hunk 9 above)
Comment (shreekarSS) [reply tests/helpers.py]: Yes, will rebase

File: tests/helpers.py (hunk 5 above)
Comment (shreekarSS) [reply tests/helpers.py]: Ya, we can directly call .get() instead of putting that in a separate function. I think it's helpful if user wants to get all pvs

File: tests/manage/test_create_storage_class_pvc.py (hunk 7 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: I'm getting pv name of pvc, I could see a method backed_pv() would do that, so modifying the code. Thanks!

File: tests/manage/test_create_storage_class_pvc.py (hunk 8 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: I'm getting pv name of pvc, I could see a method backed_pv() would do that, so modifying the code. Thanks!

File: tests/manage/test_create_storage_class_pvc.py (hunk 6 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: Made in capital letters , fixed

**Comment by shreekarSS** [comment]
> This is a same PR as #332 and I hope it hasn't changed from what was in the original PR, @shreekarSS can you address those comments fast, I would like to see it get merged fast so that we can see >90% PASS RATE for tier1.

Updated  PR with changes requested.

File: tests/helpers.py (hunk 4 above)
Comment (shreekarSS) [reply tests/helpers.py]: Updated !

File: tests/helpers.py [hunk 11]
```diff
@@ -527,3 +551,42 @@ def run_io_with_rados_bench(**kw):
...
+    ocp_pv_list = get_all_pvs()
+    pv_all_list = ocp_pv_list['items']
+    if pv_all_list:
+        for item in pv_all_list:
+            logging.info(item['metadata']['name'])
+            logging.info(pv)
+            if item['metadata']['name'] == pv:
+                raise AssertionError
+            else:
+                return True
```
Comment (clacroix12) [inline tests/helpers.py]: Do we actually want to return True here? We will end up returning True for the first `item` that doesn't raise an `AssertionError` and fail to check the subsequent items in `pv_all_list`. Seems to me that we should remove this `else: return True` and only return True after all items in `pv_all_list` have been iterated over and none have raised an `AssertionError`.

File: tests/manage/test_create_storage_class_pvc.py [hunk 12]
```diff
@@ -1,135 +1,133 @@
...
+    log.info("Deleting RBD Storageclass")
+    RBD_SC_OBJ.delete()
+
+
+@pytest.fixture(scope='function')
+def test_fixture_cephfs(request):
+    def finalizer():
+        teardown_fs()
+
+    request.addfinalizer(finalizer)
```
Comment (clacroix12) [inline tests/manage/test_create_storage_class_pvc.py]: Same here, add the finalizer directly.
```suggestion
    request.addfinalizer(teardown_fs)
```

File: tests/manage/test_create_storage_class_pvc.py [hunk 13]
```diff
@@ -1,135 +1,133 @@
...
 
 
-POD = ocp.OCP(kind='Pod', namespace=config.ENV_DATA['cluster_namespace'])
-CEPH_OBJ = None
+@pytest.fixture(scope='function')
+def test_fixture_rbd(request):
+    def finalizer():
+        teardown_rbd()
+
+    request.addfinalizer(finalizer)
```
Comment (clacroix12) [inline tests/manage/test_create_storage_class_pvc.py]: You can add `teardown_fbd` as the finalizer directly here, no need to wrap it in the `finalizer` function.
```suggestion
    request.addfinalizer(teardown_rbd)
```

**Comment by vasukulkarni** [comment]
@shreekarSS file an issue for pending comments and take it up as follow on PR, we have to merge this to get CI in good state.

**Comment by vasukulkarni** [comment]
@shreekarSS fix the CI failures and we can merge

File: tests/manage/test_create_storage_class_pvc.py (hunk 12 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: Thanks, added in the script

File: tests/manage/test_create_storage_class_pvc.py (hunk 13 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: Thanks, added in the script

**Comment by shreekarSS** [comment]
> Please adjust sure create_pod() calls to the function change

Yes,modified the files which were having create_pod() function

**Comment by vasukulkarni** [comment]
@ebenahar @petr-balogh @RazTamir can I get one quick approval here for merge, we can take up anything minor as a follow on PR, this gets us to nearly *clean* run of tier1 tests.

File: tests/helpers.py [hunk 14]
```diff
@@ -240,9 +267,14 @@ def create_pvc(sc_name, pvc_name=None):
...
     pvc_data['metadata']['namespace'] = defaults.ROOK_CLUSTER_NAMESPACE
     pvc_data['spec']['storageClassName'] = sc_name
-    return create_resource(
-        desired_status=constants.STATUS_BOUND, **pvc_data
-    )
+    if size:
+        pvc_data['spec']['resources']['requests']['storage'] = size
+    ocs_obj = pvc.PVC(**pvc_data)
+    ocs_obj.create()
+    ocs_obj.reload()
```
Comment (ebenahar) [inline tests/helpers.py]: why is this needed? create() does reload()

Comment (shreekarSS) [reply tests/helpers.py]: Yes, but 'backed_pv' was still None cus it takes little time to update 'pv' once pvc created as I have observed.

File: tests/helpers.py [hunk 15]
```diff
@@ -217,18 +243,19 @@ def create_storage_class(
         del sc_data['parameters']['userid']
     except KeyError:
         pass
-    return create_resource(**sc_data, wait=False)
+    sc_obj = create_resource(**sc_data)
```
Comment (RazTamir) [inline tests/helpers.py]: return create_resource(**sc_data)

File: tests/helpers.py [hunk 16]
```diff
@@ -57,41 +55,65 @@ def create_resource(
...
+            condition=state, resource_name=resource.name
+        )
+    except TimeoutExpiredError:
+        logging.warning(f'Failed to reach {resource.kind} {resource.name} to desired {state}')
+        return False
+    logger.info(f"{resource.kind} {resource.name} reached state {state}")
+    return True
+
+
+def create_pod(interface_type=None, pvc=None, wait=True):
```
Comment (RazTamir) [inline tests/helpers.py]: pvc_name

File: tests/helpers.py [hunk 17]
```diff
@@ -217,18 +243,19 @@ def create_storage_class(
...
     Create a PVC
 
     Args:
         sc_name (str): The name of the storage class for the PVC to be
             associated with
         pvc_name (str): The name of the PVC to create
-
+        size(str): Size of pvc to create
     Returns:
         OCS: An OCS instance for the PVC
```
Comment (RazTamir) [inline tests/helpers.py]: PVC object

File: tests/helpers.py (hunk 14 above)
Comment (RazTamir) [reply tests/helpers.py]: so do the reload() only if you pass wait=True.
If you don't want to wait this second call to reload() will take valuable time when we'll do disruptive operations

File: tests/helpers.py [hunk 18]
```diff
@@ -527,3 +553,40 @@ def run_io_with_rados_bench(**kw):
...
+
+def get_all_pvs():
+    """
+    Gets all pv in given namespace
+
+    Returns:
+         dict: Dict of all pvc in namespaces
+    """
+    ocp_pv_obj = ocp.OCP(
+        kind=constants.PV, namespace=defaults.ROOK_CLUSTER_NAMESPACE)
```
Comment (RazTamir) [inline tests/helpers.py]: ) in new line

File: tests/helpers.py [hunk 19]
```diff
@@ -527,3 +553,40 @@ def run_io_with_rados_bench(**kw):
...
     logger.info("Finished radosbench")
     return ret
+
+
+def get_all_pvs():
+    """
+    Gets all pv in given namespace
+
+    Returns:
+         dict: Dict of all pvc in namespaces
```
Comment (RazTamir) [inline tests/helpers.py]: in namespaces --> in openshift-storage namespace

File: tests/helpers.py (hunk 5 above)
Comment (RazTamir) [reply tests/helpers.py]: Sorry but I don't see the point doing that - this is time-consuming.
Please fix

File: tests/helpers.py (hunk 4 above)
Comment (RazTamir) [reply tests/helpers.py]: Why not returning False?

File: tests/manage/test_create_storage_class_pvc.py [hunk 20]
```diff
@@ -1,135 +1,129 @@
...
+        log.info('creating pvc for RBD ')
+        pvc_name = helpers.create_unique_resource_name(
+            'test-rbd', 'pvc'
         )
-        del self.cephfs_secret['data']['userID']
-        del self.cephfs_secret['data']['userKey']
-        self.cephfs_secret['data']['adminKey'] = (
-            get_admin_key_from_ceph_tools()
+        RBD_PVC_OBJ = helpers.create_pvc(
+            sc_name=RBD_SC_OBJ.name, pvc_name=pvc_name)
```
Comment (RazTamir) [inline tests/manage/test_create_storage_class_pvc.py]: ) in new line

**Comment by vasukulkarni** [comment]
@shreekarSS please address some more comments and lets gets this merged ASAP, Thanks!

File: tests/helpers.py [hunk 21]
```diff
@@ -240,9 +267,14 @@ def create_pvc(sc_name, pvc_name=None):
...
     pvc_data['spec']['storageClassName'] = sc_name
-    return create_resource(
-        desired_status=constants.STATUS_BOUND, **pvc_data
-    )
+    if size:
+        pvc_data['spec']['resources']['requests']['storage'] = size
+    ocs_obj = pvc.PVC(**pvc_data)
+    ocs_obj.create()
+    if wait:
+        ocs_obj.reload()
```
Comment (RazTamir) [inline tests/helpers.py]: maybe move the reload after the waiting for it?
Then you will get the actual status of the PVC

File: tests/helpers.py [hunk 22]
```diff
@@ -527,3 +553,41 @@ def run_io_with_rados_bench(**kw):
...
+    """
+    validates if pv is deleted after pvc deletion
+
+    Args:
+        pv_name (str): pv from pvc to validates
+    Returns:
+        bool: True if deletion is successful
+
+    Raises:
+        CommandFailed: If pv is not deleted
```
Comment (RazTamir) [inline tests/helpers.py]: Should be AssertionError

File: tests/manage/test_create_storage_class_pvc.py [hunk 23]
```diff
@@ -1,135 +1,130 @@
...
-        self.cephfs_sc = templating.load_yaml_to_dict(
-            constants.CSI_CEPHFS_STORAGECLASS_YAML
+        RBD_PVC_OBJ = helpers.create_pvc(
+            sc_name=RBD_SC_OBJ.name, pvc_name=pvc_name
         )
-        self.cephfs_sc['parameters']['monitors'] = self.mons
-        self.cephfs_sc['parameters']['pool'] = (
-            f"{self.fs_data['metadata']['name']}-data0"
+        if RBD_PVC_OBJ.backed_pv is None:
+            RBD_PVC_OBJ.reload()
```
Comment (RazTamir) [inline tests/manage/test_create_storage_class_pvc.py]: it seems like this has no effect.
if you will reload and the backed pv is still None?

File: tests/helpers.py [hunk 24]
```diff
@@ -57,24 +55,44 @@ def create_resource(
...
     )
-    if wait:
-        assert ocs_obj.ocp.wait_for_resource(
-            condition=desired_status, resource_name=resource_name
-        ), f"{ocs_obj.kind} {resource_name} failed to reach"
-        f"status {desired_status}"
     return ocs_obj
 
 
+<<<<<<< HEAD
```
Comment (RazTamir) [inline tests/helpers.py]: bad rebase

File: tests/helpers.py [hunk 25]
```diff
@@ -57,24 +55,44 @@ def create_resource(
...
+        resource.ocp.wait_for_resource(
+            condition=state, resource_name=resource.name
+        )
+    except TimeoutExpiredError:
+        return False
+    logger.info(f"{resource.kind} {resource.name} reached state {state}")
+    return True
+
+
+<<<<<<< HEAD
```
Comment (ebenahar) [inline tests/helpers.py]: Please fix the merge conflict

File: tests/helpers.py [hunk 26]
```diff
@@ -233,19 +260,20 @@ def create_storage_class(
...
 
     Args:
         sc_name (str): The name of the storage class for the PVC to be
             associated with
         pvc_name (str): The name of the PVC to create
-        wait (bool): True for wait for the VPC operation to complete, False
-            otherwise
+        size(str): Size of pvc to create
+        wait (bool): True for wait for the VPC operation to complete, False otherwise
+        size(str): Size of pvc to create
```
Comment (RazTamir) [inline tests/helpers.py]: double size

File: tests/helpers.py (hunk 21 above)
Comment (RazTamir) [reply tests/helpers.py]: please address this comment

File: tests/helpers.py (hunk 18 above)
Comment (RazTamir) [reply tests/helpers.py]: fix that please

File: tests/helpers.py (hunk 22 above)
Comment (RazTamir) [reply tests/helpers.py]: same - require fix

File: tests/manage/test_create_storage_class_pvc.py (hunk 23 above)
Comment (shreekarSS) [reply tests/manage/test_create_storage_class_pvc.py]: Nope, removed code

File: tests/helpers.py (hunk 21 above)
Comment (shreekarSS) [reply tests/helpers.py]: Yah, moved after wait_resource_state

**Comment by vasukulkarni** [comment]
using previous approvals, if there is anything lets file an issue and move forward since its more than 10 days old.
